net/dns/resolver/forwarder_test.go

```diff
@@ -240,3 +250,203 @@ func FuzzClampEDNSSize(f *testing.F) {
 		clampEDNSSize(data, maxResponseBytes)
 	})
 }
+
+func runDNSServer(tb testing.TB, response []byte, onRequest func(bool, []byte)) (port uint16) {
+	tcpResponse := make([]byte, len(response)+2)
+	binary.BigEndian.PutUint16(tcpResponse, uint16(len(response)))
+	copy(tcpResponse[2:], response)
+
+	// Repeatedly listen until we can get the same port.
+	var (
+		tcpLn *net.TCPListener
+		udpLn *net.UDPConn
+		err   error
+	)
+	for try := 0; try < 10; try++ {
+		if tcpLn != nil {
+			tcpLn.Close()
+			tcpLn = nil
+		}
+
+		tcpLn, err = net.ListenTCP("tcp4", &net.TCPAddr{
+			IP:   net.IPv4(127, 0, 0, 1),
+			Port: 0, // Choose one
+		})
+		if err != nil {
+			tb.Fatal(err)
+		}
+		udpLn, err = net.ListenUDP("udp4", &net.UDPAddr{
+			IP:   net.IPv4(127, 0, 0, 1),
+			Port: tcpLn.Addr().(*net.TCPAddr).Port,
+		})
+		if err == nil {
+			break
+		}
+
+	}
+	if tcpLn == nil || udpLn == nil {
+		tb.Fatal("failed to listen on same port for TCP/UDP")
+	}
```

Comment:
this seems like it'll be super flaky

Reply:
you also leak one listener

Reply:
FWIW, I ran this test with `gostress -count 5000` and it didn't flake once, so I can't say for sure that it's perfect, but it's hopefully not too bad? This just tries to get the same port, not any specific port, and lets the system assign a random one for TCP (which I'm assuming will be more contended).

Reply:
I think it depends on how happy the CI machines :)

Reply:
Done